src/renderer/canvas/canvas-renderer.ts

```diff
@@ -82,6 +132,17 @@ export class CanvasRenderer<N extends INodeBase, E extends IEdgeBase> extends Em
         this._render(graph);
       }, getThrottleMsFromFPS(this._settings.fps));
     }
+
+    // Change DPR from automatic to manual handling or change DPR value manually
+    if (!isNumber(previousDprValue) && isNumber(newDprValue) && newDprValue !== previousDprValue) {
```

Comment:
Nitpicking, but is `newDprValue !== previousDprValue` necessary here? It seems it's always true because `newDprValue` is a number while `previousDprValue` is not. On the other hand, `!isNumber(previousDprValue)` can be removed to allow users to change the DPR value more than once, but it does not seem like a real use case.

Reply:
Yep you are right, `newDrpValue !== previousDprValue` can be removed, but not the other check.

This is removing automatic observer for DPR. You can see on line 87 that it checks if settings doesn't have DPR, it will start automatic observer that checked for DPR from the browser.

So these lines 137 and 143 are for:
- 137: Stop the automatic observer - this can happen only if previousDPR was not a number, and the new value is defined
- 142: Start the observer again - this can happen only if previousDRP was a number, and the new value is null (this enables users to remove manual DPR handling).

Users can still change DPR manually which happens on line 127.